core/aws-core/src/main/java/software/amazon/awssdk/awscore/endpoint/AwsClientEndpointProvider.java

```diff
@@ -0,0 +1,461 @@
+        return clientEndpoint.getValue().isEndpointOverridden;
+    }
+
+    private ClientEndpoint resolveClientEndpoint(Builder builder) {
+        return OptionalUtils.firstPresent(clientEndpointFromClientOverride(builder),
+                                          () -> clientEndpointFromEnvironment(builder),
+                                          () -> clientEndpointFromServiceMetadata(builder))
+                            .orElseThrow(AwsClientEndpointProvider::failToLoadEndpointException);
+    }
+
+    private static SdkClientException failToLoadEndpointException() {
+        return SdkClientException.create("Unable to determine the default client endpoint. Enable TRACE logging on " +
+                                         AwsClientEndpointProvider.class.getName() + " for more information.");
+    }
+
+    private Optional<ClientEndpoint> clientEndpointFromClientOverride(Builder builder) {
+        Optional<ClientEndpoint> result = Optional.ofNullable(builder.clientEndpointOverride)
+                                                  .map(uri -> new ClientEndpoint(uri, true));
+        result.ifPresent(e -> log.trace(() -> "Client was configured with endpoint override: " + e.clientEndpoint));
+        return result;
+    }
+
+    private Optional<ClientEndpoint> clientEndpointFromEnvironment(Builder builder) {
+        if (builder.serviceEndpointOverrideEnvironmentVariable == null ||
+            builder.serviceEndpointOverrideSystemProperty == null ||
+            builder.serviceProfileProperty == null) {
+            Validate.isTrue(builder.serviceEndpointOverrideEnvironmentVariable == null &&
+                            builder.serviceEndpointOverrideSystemProperty == null &&
+                            builder.serviceProfileProperty == null ,
+                            "If any of the service endpoint override environment variable, system property or profile "
+                            + "property are configured, they must all be configured.");
+            log.trace(() -> "Environment was not checked for client endpoint.");
+            return Optional.empty();
+        }
+
+        return OptionalUtils.firstPresent(systemProperty(builder.serviceEndpointOverrideSystemProperty),
+                                          () -> systemProperty(GLOBAL_ENDPOINT_OVERRIDE_SYSTEM_PROPERTY),
+                                          () -> environmentVariable(builder.serviceEndpointOverrideEnvironmentVariable),
+                                          () -> environmentVariable(GLOBAL_ENDPOINT_OVERRIDE_ENVIRONMENT_VARIABLE),
+                                          () -> profileProperty(builder,
+                                                                builder.serviceProfileProperty + "."
+                                                                + ProfileProperty.ENDPOINT_URL),
+                                          () -> profileProperty(builder, ProfileProperty.ENDPOINT_URL))
```

Comment:
Did we miss adding this property in ProfileProperty, builds are failing due to this

Reply:
Oh I see its from the other part 2 PR, should be all good once we merge that into master and merge from master into this PR